topology/nhlt/nhlt-processor.c

```diff
@@ -0,0 +1,321 @@
+
+	snd_config_for_each(i, next, manifest) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "sof_manifest"))
+			man_inst = n;
+	}
+
+	/* if manifest not found, just exit */
+	if (!man_inst)
+		return 0;
+
+	ret = add_bytes_to_data_section(tplg_pp, src);
+	if (ret < 0)
+		return ret;
+
+	/* add the reference to the bytes into manifest data */
+	ret = snd_config_search(man_inst, "data", &data);
+	if (ret < 0)
+		return ret;
+
+	snd_config_for_each(i, next, data) {
+		const char *name;
+
+		n = snd_config_iterator_entry(i);
+		if (snd_config_get_string(n, &name) < 0)
+			continue;
+
+		/* item already exists */
+		if (!strcmp(name, data_name))
+			return 0;
+		id_num++;
+	}
+
+	/* add new item */
+	id_name = tplg_snprintf("%d", id_num);
+	if (!id_name)
+		return -ENOMEM;
+
+	ret = snd_config_make(&cfg_new, id_name, SND_CONFIG_TYPE_STRING);
+	free(id_name);
+	if (ret < 0)
+		return ret;
+
+	ret = snd_config_set_string(cfg_new, data_name);
+	if (ret < 0)
+		return ret;
+
+	ret = snd_config_add(data, cfg_new);
+	if (ret < 0)
+		snd_config_delete(cfg_new);
+
+	return 0;
+}
+
+static int print_nhlt_as_hex_bytes(struct nhlt *blob, struct endpoint_descriptor **eps, char *dst)
+{
+	uint8_t *top_p = (uint8_t *)blob;
+	struct endpoint_descriptor *ep;
+	uint8_t *ep_p;
+	int i, j;
+
+	for (i = 0; i < sizeof(struct nhlt); i++) {
+		snprintf(dst, 6, "0x%02x,", *top_p);
+		dst += 5;
+		top_p++;
+	}
+
+	for (i = 0; i < blob->endpoint_count; i++) {
+		ep = eps[i];
+		ep_p = (uint8_t *)ep;
+		printf("ep %i length is %u\n", i, ep->length);
+		for (j = 0; j < ep->length; j++) {
+			snprintf(dst, 6, "0x%02x,", *ep_p);
+			dst += 5;
+			ep_p++;
+		}
+	}
+
+	/* remove the last comma... */
+	dst--;
+	*dst = '\0';
+
+	return 0;
+}
+
+static bool check_for_nhlt(struct tplg_pre_processor *tplg_pp)
+{
+	snd_config_iterator_t i, next, i2, next2;
+	snd_config_t *n, *n2, *manifest;
+	const char *id;
+	const char *nhlt;
+	int ret;
+
+	/* find manifest section and its named instance */
+	ret = snd_config_search(tplg_pp->input_cfg, "Object.Base.manifest", &manifest);
+	if (ret < 0)
+		return false;
+
+	snd_config_for_each(i, next, manifest) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		snd_config_for_each(i2, next2, n) {
+
+			n2 = snd_config_iterator_entry(i2);
+
+			if (snd_config_get_id(n2, &id) < 0)
+				continue;
+
+			if (!strcmp(id, "nhlt")) {
+				if (snd_config_get_string(n2, &nhlt))
+					return false;
+
+				if(!strcmp(nhlt, "true"))
+					return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+/* called at the beginning of topology pre-processing */
+int nhlt_init(struct tplg_pre_processor *tplg_pp)
+{
+	if (!tplg_pp)
+		return -EINVAL;
+
+	return 0;
+}
+
+/* called when exiting topology pre-processing */
+int nhlt_delete(struct tplg_pre_processor *tplg_pp)
+{
+	if (!tplg_pp)
+		return -EINVAL;
+
+	if(tplg_pp->private_data)
+		free(tplg_pp->private_data);
```

Comment:
you dont seem to allocate preivate_data in init()?

Reply:
Allocated in nhlt_dmic_init_params(tplg_pp) or nhlt_ssp_init_params(tplg_pp). Not very clear, but I' trying to separate the nhlt and the blob generation (headers and code) from each other.